src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -2029,20 +2052,43 @@ private unsafe void WriteToAllListeners(int eventId, Guid* activityID, Guid* chi
                 ReportOutOfBandMessage(SR.Format(SR.EventSource_EventParametersMismatch, eventId, eventDataCount, metadata.Parameters.Length));
             }
 
-            object?[]? args;
-
+            object?[] args;
             if (eventDataCount == 0)
             {
                 args = Array.Empty<object>();
             }
             else
             {
                 ParameterInfo[] parameters = metadata.Parameters;
-                args = new object[Math.Min(eventDataCount, parameters.Length)];
-                EventSource.EventData* dataPtr = data;
-                for (int i = 0; i < args.Length; i++)
+                args = new object?[Math.Min(eventDataCount, parameters.Length)];
+
+                if (metadata.AllParametersAreString)
                 {
-                    args[i] = DecodeObject(parameters[i].ParameterType, ref dataPtr);
+                    for (int i = 0; i < args.Length; i++, data++)
+                    {
+                        IntPtr dataPointer = data->DataPointer;
+#if DEBUG
+                        Debug.Assert(data->Size % 2 == 0, "String size should be even");
+                        for (int j = 0; j < data->Size / 2 - 1; j++) Debug.Assert(*((char*)dataPointer + j) != 0, "String may not contain null chars");
+                        Debug.Assert(*((char*)dataPointer + data->Size / 2 - 1) == 0, "String must be null terminated");
+#endif
+                        args[i] = dataPointer == IntPtr.Zero ? null : new string((char*)dataPointer, 0, (data->Size >> 1) - 1);
+                    }
+                }
+                else if (metadata.AllParametersAreInt32)
+                {
+                    for (int i = 0; i < args.Length; i++, data++)
+                    {
+                        Debug.Assert(data->Size == 4);
+                        args[i] = *(int*)data->DataPointer;
+                    }
+                }
+                else
+                {
+                    for (int i = 0; i < args.Length; i++, data++)
+                    {
```

Comment:
There might be a little more perf to gain if the for loop is inside the method rather than outside of it. I'm guessing DecodeObject() isn't being inlined right now and making the loop body opaque inside the method doesn't give the JIT as much visbility to optimize. There is probably also a tiny overhead in repeated execution of the DecodeObject prolog/epilog.

Reply:
Oh nice, ~10% is even more than I would have guessed : )